sphinx/directives/__init__.py

```diff
@@ -70,18 +70,25 @@ class ObjectDescription(SphinxDirective):
     # Warning: this might be removed in future version. Don't touch this from extensions.
     _doc_field_type_map = {}  # type: Dict[str, Tuple[Field, bool]]
 
+    def _process_type_map(self, typelist):
+        # type: (List[Field]) -> Dict[str, Tuple[Field, bool]]
+        typemap = {}
+        for field in typelist:
+            for name in field.names:
+                typemap[name] = (field, False)
+            if field.is_typed:
+                typed_field = cast(TypedField, field)
+                for name in typed_field.typenames:
+                    typemap[name] = (field, True)
+        return typemap
+
+    def __init__(self, *args, **kwargs):
+        # type: (*Any, **Any) -> None
+        SphinxDirective.__init__(self, *args, **kwargs)
+        self._doc_field_type_map = self._process_type_map(self.doc_field_types)
+
     def get_field_type_map(self):
         # type: () -> Dict[str, Tuple[Field, bool]]
-        if self._doc_field_type_map == {}:
-            for field in self.doc_field_types:
```

Comment:
Could you override `self._doc_field_type_map` here please? For some reasons, I'd not like to add a new method.

Reply:
Hey @tk0miya -- I'm unsure exactly what you'd like here, but are you thinking something like this?
``` python
_doc_field_type_map = {}

def __init__ (self, *args, **kwargs):
  SphinxDirective.__init__(self, *args, **kwargs)
  self._doc_field_type_map = self.get_field_type_map(self):

def get_field_type_map(self):
  if self._doc_field_type_map == {}:
    // if unset, set self._doc_field_type_map, as the old code used to
  return self._doc_field_type_map
```

Reply:
Hey @tk0miya -- I'm sorry, I'm unsure what you mean here... do you mean to move the code I'd put into `_process_type_map` into the constructor? Or do you mean that you'd prefer for `get_field_type_map` to always rebuild the type map?

Reply:
Hey @tk0miya -- I'm sorry, I'm unsure what you mean here... do you mean to move the code I'd put into `_process_type_map` into the constructor? Or do you mean that you'd prefer for `get_field_type_map` to always rebuild the type map?

Reply:
For example, this suggested change seems lighter and more in line with the existing code, but adds an optional keyword argument to `get_field_type_map`:
``` python
    # Warning: this might be removed in future version. Don't touch this from extensions.
    _doc_field_type_map = {}  # type: Dict[str, Tuple[Field, bool]]

    def __init__(self, *args, **kwargs):
        # type: (*Any, **Any) -> None
        SphinxDirective.__init__(self, *args, **kwargs)
        self._doc_field_type_map = self.get_field_type_map(force_rebuild=True)

    def get_field_type_map(self, force_rebuild=False):
        # type: (boolean) -> Dict[str, Tuple[Field, bool]]
        if force_rebuild or self._doc_field_type_map == {}:
            for field in self.doc_field_types:
                for name in field.names:
                    self._doc_field_type_map[name] = (field, False)
                if field.is_typed:
                    typed_field = cast(TypedField, field)
                    for name in typed_field.typenames:
                        self._doc_field_type_map[name] = (field, True)
        return self._doc_field_type_map

```

Reply:
sorry for poor comment. I meant this code.
```
    def get_field_type_map(self) -> Dict[str, Tuple[Field, bool]]:
        if self._doc_field_type_map == {}:
            self._doc_field_type_map = {}    # OVERRIDE the class variable by instance's
            for field in self.doc_field_types:
                for name in field.names:
                    self._doc_field_type_map[name] = (field, False)

                if field.is_typed:
                    typed_field = cast(TypedField, field)
                    for name in typed_field.typenames:
                        self._doc_field_type_map[name] = (field, True)

        return self._doc_field_type_map
```

Reply:
Oh! I see -- yes, that is much simpler. I have pushed a new commit to change as you suggest.